cpp/src/strings/utilities.cu

```diff
@@ -147,9 +148,30 @@ std::mutex g_flags_table_mutex;
 std::mutex g_cases_table_mutex;
 std::mutex g_special_case_mappings_mutex;
 
-character_flags_table_type* d_character_codepoint_flags = nullptr;
-character_cases_table_type* d_character_cases_table     = nullptr;
-special_case_mapping* d_special_case_mappings           = nullptr;
+template <typename TableType>
+class per_context_cache {
+ public:
+  template <typename Initializer>
+  TableType* find_or_initialize(const Initializer& init)
+  {
+    CUcontext c;
+    cuCtxGetCurrent(&c);
+    auto finder = cache_.find(c);
+    if (finder == cache_.end()) {
+      TableType* result = init();
+      cache_[c]         = result;
+      return result;
+    } else
+      return finder->second;
+  }
+
+ private:
+  std::unordered_map<CUcontext, TableType*> cache_;
+};
```

Comment:
This is awesome.
Please add a bunch of comments here describing what this class does and the problem it solves.

Reply:
Ok. I'll add some comments in the following commit shortly.

Reply:
Now you will find comments there. Please take a look and resolve this if you're satisfied.